src/main/java/io/javalin/core/JavalinServlet.kt

```diff
@@ -74,7 +75,44 @@ class JavalinServlet(
                 }
             }
             if (endpointEntries.isEmpty() && type != HandlerType.HEAD && type != HandlerType.GET) {
-                throw HaltException(404, "Not found")
+
+                if (prefer405over404) {
+                    val availableHandlerTypes = ArrayList<HandlerType>()
+
+                    enumValues<HandlerType>().forEach {
+                        val entries = matcher.findEntries(it, requestUri)
+
+                        if (!entries.isEmpty()) {
+                            availableHandlerTypes.add(it)
+                        }
+                    }
+
+                    if (availableHandlerTypes.isEmpty()) {
+                        throw HaltException(404, "Not found")
+                    }
+
+                    val htmlTemplate =
+                        """
+                            <!DOCTYPE html>
+                            <html lang="en">
+                                <head>
+                                    <meta charset="UTF-8">
+                                    <title>Method Not Allowed</title>
+                                </head>
+                                <body>
+                                    <h1>405 - Method Not Allowed</h1>
+                                    <p>
+                                        Available Methods: <strong>${availableHandlerTypes.joinToString(", ")}</strong>
+                                    </p>
+                                </body>
+                            </html>
+                        """
+                    ctx.status(405)
```

Comment:
I did not study the code thoroughly, but I think it would be better to throw HaltException(405) (if possible) here instead of rendering 405 as if it would be normal. That is because I think that exception mapping won't be triggered for this https://javalin.io/documentation#exception-mapping

Here I mentioned that there might be couple of places in Javalin that are hardcoded, which may be undersirable given some strict requirements, this would be one of them.
https://github.com/tipsy/javalin/issues/219